crates/matrix-sdk-ui/src/room_list_service/filters/category.rs

```diff
@@ -0,0 +1,181 @@
+use matrix_sdk::{Client, RoomListEntry};
+
+use super::Filter;
+
+/// An enum to represent whether a room is about “people” (1 or 2 users) or
+/// “group” (more than 2 users).
+///
+/// This is implemented this way so that it's impossible to filter by “group”
+/// and by “people” at the same time: these criteria are mutually
+/// exclusive by design per filter.
+#[derive(Copy, Clone, PartialEq)]
+pub enum RoomCategory {
+    Group,
+    People,
+}
+
+type DirectTargetsLength = usize;
+
+struct CategoryRoomMatcher<F>
+where
+    F: Fn(&RoomListEntry) -> Option<DirectTargetsLength>,
+{
+    /// _Direct targets_ mean the number of users in a direct room, except us.
+    /// So if it returns 1, it means there are 2 users in the direct room.
+    number_of_direct_targets: F,
+}
+
+impl<F> CategoryRoomMatcher<F>
+where
+    F: Fn(&RoomListEntry) -> Option<DirectTargetsLength>,
+{
+    fn matches(&self, room_list_entry: &RoomListEntry, expected_kind: RoomCategory) -> bool {
+        if !matches!(room_list_entry, RoomListEntry::Filled(_) | RoomListEntry::Invalidated(_)) {
+            return false;
+        }
+
+        let kind = match (self.number_of_direct_targets)(room_list_entry) {
+            // If 1, we are sure it's a direct room between two users. It's the strict
+            // definition of the `People` category, all good.
+            Some(1) => RoomCategory::People,
+
+            // If smaller than 1, we are not sure it's a direct room, it's then a `Group`.
+            // If greater than 1, we are sure it's a direct room but not between
+            // two users, so it's a `Group` based on our expectation.
```

Comment:
I assume we know that this is totally different from Element Web's idea of whether a room is a DM?

Reply:
I totally disagree with the fact we need to count the number of participants in a room to decide whether it should be part of the People Category. I'm not doing the rule here :-(.